src/mds/MDSRank.cc

```diff
@@ -2613,11 +2614,31 @@ void MDSRankDispatcher::handle_asok_command(
   int r = 0;
   CachedStackStringStream css;
   bufferlist outbl;
-  if (command == "dump_ops_in_flight" ||
-      command == "ops") {
+  dout(10) << __func__ << ": " << command << dendl;
+  if (command == "dump_ops_in_flight") {
     if (!op_tracker.dump_ops_in_flight(f)) {
       *css << "op_tracker disabled; set mds_enable_op_tracker=true to enable";
     }
+  } else if (command == "ops") {
+    vector<string> flags;
+    cmd_getval(cmdmap, "flags", flags);
+    std::unique_lock l(mds_lock, std::defer_lock);
+    auto lambda = OpTracker::default_dumper;
+    if (flags.size()) {
+      /* use std::function if we actually want to capture flags someday */
+      lambda = [](const TrackedOp& op, Formatter* f) {
+        auto* req = dynamic_cast<const MDRequestImpl*>(&op);
+        if (req) {
+          req->dump_with_mds_lock(f);
+        } else {
+          op.dump_type(f);
+        }
+      };
+      l.lock();
+    }
+    if (!op_tracker.dump_ops_in_flight(f, false, {""}, false, lambda)) {
+      *css << "op_tracker disabled; set mds_enable_op_tracker=true to enable";
+    }
```

Comment:
```
        {
            "description": "client_request(client.4226:2033 rename #0x10000000508/config #0x10000000508/config.lock 2023-08-02T11:56:34.927444+0800 caller_uid=1000, caller_gid=1000{10,1000,})",
            "initiated_at": "2023-08-02T11:56:34.929247+0800",
            "age": 0.21241235,
            "duration": 0.21261556200000001,
            "type_data": {
                "flag_point": "acquired locks",
                "reqid": {
                    "entity": {
                        "type": "client",
                        "num": 4226
                    },
                    "tid": 2033
                },
                "op_type": "client_request",
                "events": [
                    {
                        "time": "2023-08-02T11:56:34.929247+0800",
                        "event": "initiated"
                    },
                    {
                        "time": "2023-08-02T11:56:34.929247+0800",
                        "event": "header_read"
                    },
                    {
                        "time": "2023-08-02T11:56:34.929248+0800",
                        "event": "throttled"
                    },
                    {
                        "time": "2023-08-02T11:56:34.929255+0800",
                        "event": "all_read"
                    },
                    {
                        "time": "2023-08-02T11:56:34.930311+0800",
                        "event": "dispatched"
                    },
                    {
                        "time": "2023-08-02T11:56:34.930631+0800",
                        "event": "acquired locks"
                    }
                ],
                "locks": null
            }

```

Why the `locks` still `null` after acquiring the locks ?

Did I miss something @batrick ?

Reply:
What command did you run?

Reply:
> What command did you run?

`./bin/ceph daemon mds.a dump cache`.

Reply:
Huh, you got an op dump out of `dump cache`?

Reply:
Sorry, my bad. 

It's `./bin/ceph daemon mds.a ops`.

Reply:
I know you just leave the `locks` to be `null` always for `dump_ops_in_flight`, so I tried the `ops` instead by running some tests, like `git clone ...`.

From my understanding when an OP acquires any lock, in the `locks` we should see it, right ?

Reply:
No, you have to ask for it. That code path does not have the mds_lock already. It only acquires it if you want the detailed `ops` dump (i.e. `flags="locks"`).

Reply:
> You need to run now:
> 
> `bin/ceph daemon mds.a ops locks`
> 

Okay, this works. Sorry I missed this parameter.

> It's not automatic to avoid locking the `mds_lock` if unnecessary. Also, if there is a bug we don't want to routinely trigger it if some hypothetical monitoring tool is using `ops` to introspect on the MDS.

Yeah, with the `locks` parameter it will generate a lot more outputs.